docs(oidc): web keys (#8508)
zitadel/zitadel: docs/docs/guides/integrate/login/oidc/webkeys.md

---
title: OpenID Connect and Oauth2 web keys
sidebar_label: Web keys
---

> Web Keys in ZITADEL are used to sign and verify JSON Web Tokens (JWT).
> ID tokens are created, signed and returned by ZITADEL when a OpenID connect (OIDC) or Oauth2
> authorization flow completes and a user is authenticated.
> Optionally zitadel can return JWTs for access tokens if the OIDC Application is configured for it.

Reviewer: ```suggestion
Web Keys in ZITADEL are used to sign and verify JSON Web Tokens (JWT).
ID tokens are created, signed and returned by ZITADEL when a OpenID connect (OIDC) or OAuth2
authorization flow completes and a user is authenticated.
Optionally, ZITADEL can return JWTs for access tokens if the OIDC Application is configured for it.
```

## Introduction

ZITADEL uses asymmetric cryptography to sign and validate JWTs.
Keys are generated in pairs resulting in a private and public key.
Private keys are used to sign tokens.
Public keys are used to verify tokens.
OIDC clients need the public key to verify ID tokens.

> Oauth2 API apps might need the public key if they want to client-side verification of a
> JWT access tokens, instead of [introspection](/docs/apis/openidoauth/endpoints#introspection_endpoint).
> ZITADEL uses public key verification when API calls are made or when the user-info or introspection

Reviewer: ```suggestion
OAuth2 API apps might need the public key if they want to client-side verification of a
JWT access tokens, instead of [introspection](/docs/apis/openidoauth/endpoints#introspection_endpoint).
ZITADEL uses public key verification when API calls are made or when the userInfo or introspection
```

:::

### JSON Web Key

ZITADEL implements the [RFC7517 - JSON Web Key (JWK)](https://www.rfc-editor.org/rfc/rfc7517) format for storage and distribution of public keys.
Web keys in ZITADEL support a number of [JSON Web Algorithms (JWA)](https://www.rfc-editor.org/rfc/rfc7518) for digital signatures:

| Identifier | Description                     |
| ---------- | ------------------------------- |
| RS256      | RSASSA-PKCS1-v1_5 using SHA-256 |
| RS384      | RSASSA-PKCS1-v1_5 using SHA-384 |
| RS512      | RSASSA-PKCS1-v1_5 using SHA-512 |
| ES256      | ECDSA using P-256 and SHA-256   |
| ES384      | ECDSA using P-384 and SHA-384   |
| ES512      | ECDSA using P-521 and SHA-512   |
| EdDSA      | EdDSA signature algorithms[^1]  |

[^1]: EdDSA refers to both Ed25519 and Ed448 curves. ZITADEL only supports Ed25519 with a SHA-512 hashing algorithm. EdDSA is for JSON Object Signing is defined in [RFC8037](https://www.rfc-editor.org/rfc/rfc8037).

### Client Algorithm Support

> Before customizing the algorithm the instance admin **MUST** make sure the complete App and API ecosystem

Reviewer: ```suggestion
Before customizing the algorithm the instance admin **MUST** make sure the complete app and API ecosystem
```

supports the chosen algorithm.

When all OIDC applications of an instance use opaque access tokens, and they call APIs which only use
introspection for token validation, only the OIDC applications will need to support the chosen algorithm.
If JWT access tokens are used and APIs do public key verification, those APIs need to support the chosen algorithm as well.

RS256 is widely considered the default algorithm and must be supported by all OIDC/Oauth2 providers, relying parties and resource servers.
This is also the default ZITADEL uses when [creating web keys](#creation).
It might be reasonable to assume RS384 and RS512 are just as supported, because those are just variations on RSA based keys.
The ES256, ES384 and ES512 might have reasonable support as well,
ECDSA is part of the same [JSON Web Algorithms (JWA)](https://www.rfc-editor.org/rfc/rfc7518) as RSA.

EdDSA usage is defined in the supplemental [RFC8037](https://www.rfc-editor.org/rfc/rfc8037),

> and therefore may be less supported then the others.

Reviewer: ```suggestion
and therefore may be less supported than the others.
```

Since the decision in the proposal isn't published yet,
there is a big change some OIDC client libraries don't have proper support for EdDSA / ed25519.

The ZITADEL back-end is written in Go. The Go developers have denied ed448 curve implementations to be included.
Therefore ZITADEL only uses ed25519 with a SHA512.
The same counts for [zitadel/oidc](https://github.com/zitadel/oidc) Go library.

## Web Key management

ZITADEL provides a resource based [web keys API](/docs/apis/resources/webkey_service_v3).
The API allows the creation, activation, deletion and listing of web keys.
All public keys that are stored for an instance are served on the [JWKS endpoint](#json-web-key-set).
Applications need public keys for token verification and not all applications are capable of on-demand
key fetching when receiving a token with an unknown key ID (`kid` header claim).
Instead, those application may do a time-based refresh or only load keys at startup.

Using the web keys API, keys can be created first and activated for signing later.
This allows the keys to be distributed to the instance's apps and caches.

> Once a key is deactivated, its public key will remain available for token verification on the web key is deleted.

Reviewer: ```suggestion
Once a key is deactivated, its public key will remain available for token verification until the web key is deleted.
```

Delayed deletion makes sure tokens that were signed before the key got deactivated remain valid.

When the `web_key` [feature](/docs/apis/resources/feature_service_v2/feature-service-set-instance-features) is enabled the first time,
two web key pairs are created with one activated.

### Creation

The web key [create](/docs/apis/resources/webkey_service_v3/zitadel-web-keys-create-web-key) endpoint generates a new web key pair,
using the passed generator configuration from the request. This config is a one-of field of:

- RSA
- ECDSA
- ED25519

When the request does not contain any specific configuration,
[RSA](#rsa) is used as default with the default options as described below:

```bash

> curl -L 'https://$CUSTOM-DOMAIN/resources/v3alpha/web_keys/' \

Reviewer: ```suggestion
curl -L 'https://$CUSTOM-DOMAIN/resources/v3alpha/web_keys' \
```

-H 'Content-Type: application/json' \
-H 'Accept: application/json' \
-H 'Authorization: Bearer <TOKEN>' \
-d '{}'
```

#### RSA

The RSA generator config takes two enum values.

- The `bits` fields determines the size of the RSA key:
  - `RSA_BITS_2048` (**default**)
  - `RSA_BITS_3072`
  - `RSA_BITS_4096`
- The `hasher` field sets the hash mode and
  determines the `alg` header value of the web key:
  - `RSA_HASHER_SHA256` results in the RS256 algorithm header. (**default**)
  - `RSA_HASHER_SHA384` results in the RS384 algorithm header.
  - `RSA_HASHER_SHA512` results in the RS512 algorithm header.

For example, to create a RSA web key with the size of 3072 bits and the SHA512 algorithm (RS512):

```bash

> curl -L 'https://$CUSTOM-DOMAIN/resources/v3alpha/web_keys/' \

Reviewer: ```suggestion
curl -L 'https://$CUSTOM-DOMAIN/resources/v3alpha/web_keys' \
```

-H 'Content-Type: application/json' \
-H 'Accept: application/json' \
-H 'Authorization: Bearer <TOKEN>' \
-d '{
  "rsa": {
    "bits": "RSA_BITS_3072",
    "hasher": "RSA_HASHER_SHA512"
  }
}'

```

#### ECDSA

The ECDSA generator config takes a single `curve` enum value which determines both the key's curve parameters and hashing algorithm:

- `ECDSA_CURVE_P256` uses the NIST P-256 curve and sets the ES256 algorithm header.
- `ECDSA_CURVE_P384` uses the NIST P-384 curve and sets the ES384 algorithm header.
- `ECDSA_CURVE_P512` uses the NIST P-512 curve and sets the ES512 algorithm header.

For example, to create a ECDSA web key with a P-256 curve and the SHA256 algorithm:

```bash

> curl -L 'https://$CUSTOM-DOMAIN/resources/v3alpha/web_keys/' \

Reviewer: ```suggestion
curl -L 'https://$CUSTOM-DOMAIN/resources/v3alpha/web_keys' \
```

-H 'Content-Type: application/json' \
-H 'Accept: application/json' \
-H 'Authorization: Bearer <TOKEN>' \
-d '{
  "ecdsa": {
    "curve": "ECDSA_CURVE_P256"
  }
}'
```

#### ED25519

ED25519 is an EdDSA curve and currently the only EdDSA curve supported by ZITADEL.[^2]
No config is needed for ed25519 as its specification already includes the curve parameters.
ed25519 always uses the SHA512 hasher.

Note that the `alg` header for ed25519 is `EdDSA` and refers to both ed25519 and ed448 curves.
Both curves specify different hashers.
Clients which support both curves must inspect `crv` header value to assert the difference.

For example, to create a ed25519 web key:

```bash

> curl -L 'https://$CUSTOM-DOMAIN/resources/v3alpha/web_keys/' \

Reviewer: ```suggestion
curl -L 'https://$CUSTOM-DOMAIN/resources/v3alpha/web_keys' \
```

4. Web keys are kept long enough to to allow usage of the `id_token_hint` for at least 3 months.
  Users that haven't logged in / refreshed tokens with the client app for that period,
  will need to re-enter their username.

When the feature flag was enabled the first time, the instance got two keys with the first one activated. When this feature becomes general available, instance creation will setup the first two keys in the same way. So the initial state always looks like this:

| id  | created    | changed    | state           |
| --- | ---------- | ---------- | --------------- |
| 1   | 2025-01-01 | 2025-01-01 | `STATE_ACTIVE`  |
| 2   | 2025-01-01 | 2025-01-01 | `STATE_INITIAL` |

For the sake of this example we will use simplified IDs and restrict timestamps to dates.

After one month, on 2025-02-01, we wish to activate the next available key and create a new key to be available for activation next month. This fulfills requirements 1 and 2.

```bash
curl -L -X POST 'https://$CUSTOM-DOMAIN/resources/v3alpha/web_keys/2/_activate' \
-H 'Accept: application/json' \
-H 'Authorization: Bearer <TOKEN>'

> curl -L 'https://$CUSTOM-DOMAIN/resources/v3alpha/web_keys/' \

Reviewer: ```suggestion
curl -L 'https://$CUSTOM-DOMAIN/resources/v3alpha/web_keys' \
```
